### PR TITLE
209 load historic

### PR DESCRIPTION
**Pull request by jwestw**
## Pull Request submission

This allows the user to load historic data into the pipeline. The historic data loader returns the historic data in dataframes which are stored in a dictionary. They keys to the dictionary indicate which data file the dataframe came from, e.g. "BERD202012" - see image in this comment: https://github.com/ONSdigital/research-and-development/pull/43#discussion_r1254065254

Currently all setting for this are in the developer config - but they can be moved to the userconfig easily. 

The original requirements for the ticket were: 

- Load the data
- Check that data is present (reusing functions if we can, e.g. hdfs-load-json)
- Log that data loading was successful or raise error if not  

I believe all of that is achieved and more. 

In this PR I also streamlined the importation of paths according to whether they are HDFS paths or local network drive paths. 

- To review this please run it on local Python and on Dev/Test. 
    - Change `network_or_hdfs` setting to either `network` or `hdfs`
- Try changing the years settings in the developer config. 
- All data should already be in place for you. 

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes # [RDRP-209](https://jira.ons.gov.uk/browse/RDRP-209)

#### Code

- [x] **Code runs** The code runs on my machine and/or CDSW
- [x] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [x] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [x] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [x] **Clean Code**
    - [x] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [x] Code passess flake8 linting check
    - [x] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [x] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [x] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [x] Includes `Args` and `returns` for all major functions
    - [x] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `environment.yaml`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [x] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/developer_config.yaml [hunk 1]
```diff
@@ -4,21 +4,29 @@ global:
...
+  masterlist_path: "data/external/ONSPD_NOV_2022_UK.csv"
   postcode_masterlist: "data/external/ONSPD_NOV_2022_UK.csv"
+  history_path: None
 network_paths:
+  root: None
   logs_foldername: "logs/run_logs"
   snapshot_path: "R:/BERD Results System Development 2023/DAP_emulation/spp_snapshots/snapshot-202012-002-fba5c4ba-fb8c-4a62-87bb-66c725eea5fd.json"
   postcode_masterlist: "R:/BERD Results System Development 2023/DAP_emulation/ONS_Postcode_Reference/ONSPD_NOV_2022_UK.csv"
-
+  history_path: 'R:\BERD Results System Development 2023\DAP_emulation\backseries\v1'
```
Comment (ali-i3) [inline src/developer_config.yaml]: This path points to a different folder than expected. I ran it successfully using this path:
'R:/BERD Results System Development 2023/DAP_emulation/BERD_V7_Anonymised/'

File: src/data_ingest/history_loader.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    """Creates a list of paths to load historic data from"""
+    # Create a list of paths to load
+    hist_paths = []
+    for year in history_years:
+        hist_path = hist_folder + "qv_BERD_" + str(year) + "12_qv6_reformatted.csv"
+        hist_paths.append(hist_path)
+    return hist_paths
+
+
+def load_history(year_generator: Generator[int, None, None], hist_folder_path: str, read_csv_func: callable) -> None:
```
Comment (ali-i3) [inline src/data_ingest/history_loader.py]: When you ran the pipeline, it logged which years were loaded. My run just outputted a message that showed that loading was complete. I didn't see how many files were loaded and which ones they were. Am i missing something?

File: src/data_processing/spp_snapshot_processing.py~HEAD
Comment (ali-i3) [inline src/data_processing/spp_snapshot_processing.py~HEAD:1]: Not sure why this duplicate HEAD file is getting committed to the branch. Maybe needs to be a .py~HEAD filter added to the pre-commit if it happens often

**Review by ali-i3 — COMMENTED** [review]
Looks great! I added some comments on my findings. Main issue I had was the history_path not working for me. It worked fine when you demonstrated it on your system earlier so we might have to sync paths. The pipeline.py and all the tests look good as well.

File: src/developer_config.yaml (hunk 1 above)
Comment (jwestw) [reply src/developer_config.yaml]: Think I have fixed this here: https://github.com/ONSdigital/research-and-development/pull/43/commits/c6164128e9cba2930b7a0864d13c7b2bed933bdf

File: src/data_ingest/history_loader.py (hunk 2 above)
Comment (jwestw) [reply src/data_ingest/history_loader.py]: I get this when I run the pipeline. 

image


Do you think there's a need to log anything else?

File: src/data_processing/spp_snapshot_processing.py~HEAD
Comment (jwestw) [reply src/data_processing/spp_snapshot_processing.py~HEAD:1]: I didn't realise this had been committed but yes, I keep getting this file come appear.

File: src/developer_config.yaml (hunk 1 above)
Comment (jwestw) [reply src/developer_config.yaml]: I have fixed this now.

File: src/data_ingest/history_loader.py (hunk 2 above)
Comment (jwestw) [reply src/data_ingest/history_loader.py]: Yes I have pushed that change now so you should get the same.

File: src/data_processing/spp_snapshot_processing.py~HEAD [hunk 3]
```diff
@@ -0,0 +1,64 @@
+from src.utils.helpers import Config_settings
```
Comment (AnneONS) [inline src/data_processing/spp_snapshot_processing.py~HEAD]: @jwestw  I'm curious about this ~HEAD file which is pretty different to spp_snapshot_processing. Does it mean this is the new version?

File: src/pipeline.py [hunk 4]
```diff
@@ -46,6 +47,9 @@
     MainLogger.error("The network_or_hdfs configuration is wrong")
     raise ImportError
 
+# Conditionally load paths
+paths = config[f"{network_or_hdfs}_paths"]
+
```
Comment (AnneONS) [inline src/pipeline.py]: Yes, this is an improvement

File: src/pipeline.py [hunk 5]
```diff
@@ -65,11 +69,23 @@ def run_pipeline(start):
...
     MainLogger.info("Starting Data Ingest...")
-    # Load SPP data from DAP
 
-    snapshot_path = config[f"{network_or_hdfs}_paths"]["snapshot_path"]
+    # Load historic data
+    curent_year = config["years"]["current_year"]
+    years_to_load = config["years"]["previous_years_to_load"]
+    years_gen = history_loader.history_years(curent_year, years_to_load)
+    history_path = paths["history_path"]
+    dict_of_hist_dfs = history_loader.load_history(years_gen, history_path, read_csv)
```
Comment (AnneONS) [inline src/pipeline.py]: This is currently failing for me on hdfs:
FileNotFoundException: File does not exist: /user/cdsw/None/qv_BERD_202112_qv6_reformatted.csv

What do you have in your version of HUE ?

**Review by AnneONS — COMMENTED** [review]
This runs fine on the network and the tests (excluding hdfs_mods) pass.

On hdfs the pipeline doesn't run as I get this error:
File does not exist: /user/cdsw/None/qv_BERD_202112_qv6_reformatted.csv

Also, could you rebase? My pytests fail in hdfs_mods as we don't have the latest version of develop.

I have only looked quickly through the new code so far, it looks good but I haven't got my head round all of it yet :-)

File: src/data_processing/spp_snapshot_processing.py~HEAD (hunk 3 above)
Comment (jwestw) [reply src/data_processing/spp_snapshot_processing.py~HEAD]: Oops this has been accidentally added to the PR. I will remove it

File: src/pipeline.py (hunk 5 above)
Comment (jwestw) [reply src/pipeline.py]: It's the "None" part of this path which is failing. This comes from the developer config.

Comment (jwestw) [reply src/pipeline.py]: The `history_path` should be `/ons/rdbe_dev/BERD_V7_Anonymised` I think

Comment (jwestw) [reply src/pipeline.py]: I have fixed this now and will test in dev test

File: src/data_processing/spp_snapshot_processing.py~HEAD
Comment (jwestw) [reply src/data_processing/spp_snapshot_processing.py~HEAD:1]: I have added a "~HEAD" filter to `gitignore`

**Comment by AnneONS** [comment]
Hi James, it now runs on hdfs, which is great.

However, still the pytest error on TestCheckFileExists as this is still Roddy's old version- if you rebase/ pull in develop that will solve the problem.

**Review by AnneONS — COMMENTED** [review]
This now runs in hdfs

@ali-i3  can you please check the tests, I haven't had time and there are some things I'm not familar with!

**Review by ali-i3 — APPROVED** [review]
Looks good!

**Pull request merged**